api/src/main/java/org/openmrs/api/TagService.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api;
+
+import org.openmrs.OpenmrsObject;
+import org.openmrs.TagConstants;
+import org.openmrs.annotation.Authorized;
+import org.openmrs.api.OpenmrsService;
+import org.openmrs.Tag;
+import org.openmrs.api.context.Context;
+
+import java.util.List;
+
+/**
+ * Contains services for adding/removing/obtaining Tags. Example Usage: <br>
+ * <code>
+ *   List&lt;Tag&gt; tags = Context.getService(TagService.class).getAllTags();
+ * </code>
+ * 
+ * @see org.openmrs.Tag
+ */
+
+public interface TagService extends OpenmrsService {
+	
+	/**
+	 * Returns a tag by uuid. It can be called by any authenticated user. It is fetched in read only
+	 * transaction.
+	 * 
+	 * @param uuid
+	 * @return tag or null
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	Tag getTagByUuid(String uuid);
+	
+	/**
+	 * Saves a tag. Sets the owner to superuser, if it is not set. It can be called by users with
+	 * this module's privilege. It is executed in a transaction. Tags can only be added, not edited.
+	 * 
+	 * @param tag
+	 * @return Tag that has been saved
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	Tag saveTag(Tag tag);
+	
+	/**
+	 * Completely delete the tag from the database.
+	 * 
+	 * @param tag The Tag to remove from the system
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	void deleteTag(Tag tag);
+	
+	/**
+	 * Removes a tag(if exists) from an OpenmrsObject.
+	 * 
+	 * @param openmrsObject the openmrsObject from which the tags is to be removed
+	 * @param tag the textual tag to be removed
+	 */
+	@Authorized(TagConstants.MANAGE_TAGS)
+	void removeTag(OpenmrsObject openmrsObject, String tag);
+	
+	/**
+	 * Returns a tag by id.
+	 * 
+	 * @param tagId the id of the Tag Object to be retrieved
+	 * @return Tag with given internal identifier
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	Tag getTag(Integer tagId);
+	
+	/**
+	 * Returns a list of Tag objects having the given tag, or having tags containing the substring.
+	 * Partial Search supported
+	 * 
+	 * @param tag the tag (String) which the
+	 * @return List of Tag Objects
+	 */
+	@Authorized(TagConstants.GET_TAGS)
+	List<Tag> getTags(String tag);
```

Comment:
Since this is a partial search, the parameter name should be searchPhrase for clarity